### PR TITLE
Issue13 transitive dependencies

### PR DESCRIPTION
**Pull request by jamesmcc**
I think this solves the issue, happy to make more changes if this doesnt hit the mark

### CODE REVIEW

**Comment by jerriep** [comment]
@jamesmcc Thank you for this one. 

I think this one will need some more thinking about how to approach it best. Currently, it pulls the transitive dependencies for all target frameworks which is not the correct behaviour.

As an example, here is the output when I run it for the **dotnet-outdated** project:

image

As a reference, here are the dependencies as listed on the NuGet website:

image

Since the TargetFramework we are working with is **.NETCoreApp,Version=v2.1**, it means that it is pulling in the **.NETStandard 2.0** version, and therefor it should actually only list **System.ComponentModel.Annotations 4.4.1** as a transive dependency.

This becomes tricky, since we now need to start understanding which version is the correct set of dependencies to pick based on the TargetFramework we are working with. 

Maybe the NuGet libraries has something which already does this logic for us? Not sure?

As I mentioned on the comments on #13, I think it may be better to just piggyback on some of the work MSBuild is already doing. But I am not sure whether that will work.

I need to spend time looking at this one in detail. 

You're free to carry on and have a look and see what you can come up with in the meantime...

**Comment by jamesmcc** [comment]
the dependency group holds the target framework information, I will have a look

**Comment by jerriep** [comment]
You may have to look at the [NuGet.Frameworks](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Frameworks) library to handle for you...

File: src/DotNetOutdated/Program.cs [hunk 1]
```diff
@@ -171,6 +170,22 @@ private static void WriteTargetFramework(IConsole console, Project.TargetFramewo
                 console.Write(")");
             }
             console.WriteLine();
+
+            if (this.Transitive)
```
Comment (jerriep) [inline src/DotNetOutdated/Program.cs]: please remove `this`. It is inconsitent with the rest of the code base.

File: src/DotNetOutdated/Services/NuGetPackageInfoService.cs [hunk 2]
```diff
@@ -46,18 +48,47 @@ public NuGetPackageInfoService()
...
 
                 allVersions.AddRange(await findPackageById.GetAllVersionsAsync(package, _context, _logger, CancellationToken.None));
             }
 
             return allVersions;
         }
 
+        public async Task<IEnumerable<PackageDependency>> GetDependencies(string package, List<Uri> sources, NuGetVersion referencedVersion, NuGetFramework targetFramework)
+        {
+            var allDependancies = new List<PackageDependency>();
```
Comment (jerriep) [inline src/DotNetOutdated/Services/NuGetPackageInfoService.cs]: Fix spelling. Should be `allDependencies`

File: src/DotNetOutdated/Program.cs [hunk 3]
```diff
@@ -42,7 +42,12 @@ class Program : CommandBase
                                                              "Possible values: None (default), Major or Minor.",
             ShortName = "vl", LongName = "version-lock")]
         public VersionLock VersionLock { get; set; } = VersionLock.None;
-        
+
+        [Option(CommandOptionType.SingleValue, Description = "Specifies whether it should detect transitive dependencies." +
```
[the remark this answers is not in the document]
Comment (jamesmcc) [reply src/DotNetOutdated/Program.cs]: i was thinking that in the future you could pass in the depth you wanted to go, but I have made the changes as have requested

**Comment by jamesmcc** [comment]
no problem thanks for feedback, time for me to go to sleep as well

**Comment by jerriep** [comment]
@jamesmcc This is starting to look good. One more problem though...

The idea behind being able to view the transitive dependencies is so that you can manually install newer versions of this dependencies. Let's look at an example of this using the dependencies tree in VS.

As you can see, the **DotNetOutdated** project takes a dependency on **McMaster.Extensions.CommandLineUtils** (version 2.2.4). This package, in turn, has a dependency on **System.ComponentModel.Annotations** >= 4.4.1. 

Because I had not specifically added a newer version of **System.ComponentModel.Annotations** to my application, the version which my application will reference is that minimum required version - 4.4.1

image

Now, let's say that I am aware that there is a newer version (4.5.0) **System.ComponentModel.Annotations** available, I decide that I rather want that to reference that newer version and manually install 4.5.0 to my project.

This means that the version number of the transitive dependency which **McMaster.Extensions.CommandLineUtils** has should not be the minimum version required, but rather the one which is actually referenced in my project, e.g.

image

So in other words, when you report the version numbers of the transitive dependencies, you need to take this into account. Currently, it does not:

image

So in the example above, the referenced version of **System.ComponentModel.Annotations** listed under the dependencies of **McMaster.Extensions.CommandLineUtils** should have been reported as **4.5.0**.

Hope my explanation makes sense :) Thanks for your effort so far!

**Comment by jamesmcc** [comment]
I think that is handling this issue now, should we display this overwriting dependency somehow ? I am a bit torn about this but came to the decision of no.

File: src/DotNetOutdated/Program.cs [hunk 4]
```diff
@@ -83,61 +87,57 @@ public static string GetVersion() => typeof(Program)
             _projectAnalysisService = projectAnalysisService;
             _projectDiscoveryService = projectDiscoveryService;
         }
-        
+
         public async Task<int> OnExecute(CommandLineApplication app, IConsole console)
         {
             try
             {
+                Path = @"D:\OSS\fork\dotnet-outdated";
```
Comment (jerriep) [inline src/DotNetOutdated/Program.cs]: Remove this line

File: src/DotNetOutdated/Program.cs [hunk 5]
```diff
@@ -171,6 +176,22 @@ private static void WriteTargetFramework(IConsole console, Project.TargetFramewo
                 console.Write(")");
             }
             console.WriteLine();
+
+            if (!Transitive.HasValue)
```
Comment (jerriep) [inline src/DotNetOutdated/Program.cs]: Correct this condition. Should be `if (Transitive.HasValue)`

**Comment by jerriep** [comment]
> should we display this overwriting dependency somehow

The current behaviour is correct. We are technically not overriding the dependency. We are in fact displaying the correct dependency as it should be. The current behaviour is exactly what NuGet, Visual Studio, Rider, etc will do as well

**Comment by jamesmcc** [comment]
I was being lazy by not using the properties, wrists slapped.

**Comment by jerriep** [comment]
No wrist slapping. Was just wanting to make sure you know you can pass parameters from the IDE. Much easier.

**Comment by jerriep** [comment]
@jamesmcc Thanks again for this one. Just a quick note... In future when you pick up an issue, especially one like this that can involved a bit of work, just leave a comment on the issue to make sure no-one else has picked it up yet.  It is much better to just let everyone know that you are working on it, so 2 or more people do not work on the same issue at the same time.

**Comment by jamesmcc** [comment]
will do I am new to this github thing, so thanks for your patience

**Comment by jerriep** [comment]
@jamesmcc Version 1.1.0 is available on NuGet and it adds this feature. I reworked how things are done, however. Thanks again for the PR!
